### DISCUSSION TITLE
Should `createComputed` return a `Computed`? Or should `ReadSignal`  be implemented slightly differently?

### DISCUSSION THREAD

**Discussion by manuel-plavsic** [nank1ro/solidart#44] [General]
I think that the current state of `createComputed` is not ideal, since it might be slightly unclear how  `ReadSignal` (the type returned by `createComputed`) works under the hood.

In the following example,

```dart
// creating a computed
final derived = createComputed(
  // ...
);

// ...

// using the previously created computed
derived.previousValue;
```

If someone right clicks on `derived.previousValue`, the IDE will bring them to `ReadSignal`'s `previousValue`, i.e.:

```dart
@override
T? get previousValue {
  // no-op
  return null;
}
```

which will not apply since it is explicitly overridden in the `Computed` class. The same applies for other `ReadSignal` members.

I propose either:
-  returning `Computed`  instead of `ReadSingal` (with some potential changes to `Computed`)
- changing  `ReadSignal` internals so that it no longer contains a `final T _value`, but it contains a private setter that can update `_value` and `previousValue`. This way, the code for `previousValue`, `hasPreviousValue` and other affected members could be reused in `Signal`, `Computed`, ... and when the user control clicks on these members the code displayed will be valid in case of a `Computed`.

Do you like one of these 2 ideas?

**Comment by nank1ro** [comment]
I understand that this may be an inconvenience that the IDE brings to ReadSignal.
Is correct that `createComputed` returns a `ReadSignal `since the value depends on other signals and the user should not be able to manually change the value.

The logic is a bit chaotic, which is why the new reactive system is released as a dev version.

I will try to review the logic to see if it can be improved, also to make it work properly with the IDE. 
I will check if the second solution you proposed is feasible

**Comment by nank1ro** [comment]
Hi @manuel-plavsic can you check and try https://github.com/nank1ro/solidart/pull/45?
I did all the fixes and now seems working as expected. I also simplified the logic so it should work fine also with the IDE.

**Comment by manuel-plavsic** [comment]
I really like the new logic! It is so much more intuitive. Thanks!
